### DISCUSSION TITLE
my feedback - game engine for 4x game

### DISCUSSION THREAD

**Discussion by metablaster** [nem0/LumixEngine#1355] [Q&A] [answered]
I'm looking for modern game engine, it doesn't have to be complete product, I only care about:
1. good starting point for 4x game
2. good code design
3. cutting edge language and DX features

Not a single line of comment in code, **this is so big minus**, not even project readme tells anything about the engine!
Can you please at least tell if, do you recommend this engine for 4x games and how much of change to code would I need to do?

**Comment by nem0** [comment, marked as answer]
Hi, thanks for your feedback. 
1. I'm not sure what would make a game engine good starting point for 4x. This is just generic game engine, something like unity or godot (minus their scope).
2. Good code design is subjective term. Since I wrote the engine I think it has good design, why some other people think it has terrible design.
3. I'm not sure what exactly should cuttine edge language mean.

> Not a single line of comment in code

That's not true, there's this line https://github.com/nem0/LumixEngine/blob/master/src/renderer/pipeline.cpp#L403. :trollface: . You are right there are almost no comments. If you like comments I recommend RBFX, it's nicely commented as you can see [here](https://github.com/rokups/rbfx/blob/0eb40740d5797603fab97187f838202913df6a68/Source/Urho3D/Core/Spline.h#L52). 
I would probably not recommend this engine for a game. Truth is if you want to make game, use something like Unreal or Unity or some other widely used engines. On the other hand, if you want to work on an engine, this engine is very good candidate.

**Reply by metablaster** [reply]
I'm sorry if my query sounds negative, I'll try to expand:

> On the other hand, if you want to work on an engine, this engine is very good candidate.

That's exactly what I'm looking for, a minimal starting point to save my self time going from scratch, engines such as those you recommended (unity, UE) are not something we can easily modify, especially not Unreal, which is not suited for 4X at all.
And I prefer C++ over other languages.

> I'm not sure what would make a game engine good starting point for 4x.

4X is a game genre, ex. TBS (Turn Based Strategy) game, if you google out you'll see many people ask similar question, but nobody has an answer simply because game engines used for 4X game development are not publicly available or are not afodable for indie devs, there is no other option but to start from zero, which I would like to avoid by having some starting point, ex. at least an extensible game editor in addition to rendering engine if nothing else. to avoid gluing parts (audio/physics/renderer etc...) together from zero.
It's just too much of a work.

> Good code design is subjective term. Since I wrote the engine I think it has good design, why some other people think it has terrible design.

You are right, honesty I didn't even look at LumixEngine code (not yet) so I don't know and don't want to judge any ones work, my point with "good code design" would satisfy following:

1. follows known community established style guidelines and practices, consistent and popular naming convention etc...
3. compiles clean with minimum or no warnings or clean code analysis result (most devs don't like this, but they are wrong)
4. API is, documented, ex. each function/variable/class etc.. contains a solid code comment (ex. doxygen style)
5. and also already mentioned point 3. cutting edge language and DX features

> I'm not sure what exactly should cuttine edge language mean.

For example: C++17 and DX 11 (or never).

In other words, not a vintage code (which is how Stroustrup refers to old code design and practices), which is so common on GitHub and other places, because very few people upgrade their code bases, reason being usually because "it works just fine"

> You are right there are almost no comments. If you like comments I recommend RBFX, it's nicely commented as you can see..

I said lack of comment is a big minus and I'm not afraid to say this again even if that will sound negative, it's not that I care about comments being present, but I mean if API isn't documented then it's almost of no use (except if one has no other choice)

Take for example Win32 functions, each one is so well documented it's de facto a standard for how code should be documented either in sources with code comments or online on a web page.

BTW. I'm searching GitHub topics and tags for some time already and stubled upon LumixEngine which looks very promising, but I was severely disappointed for a lack of API documentation (code comments), there are many candidates and I'm just trying to find one that would meet my expectations (which I know are too strict/high).

**Reply by nem0** [reply]
Don't worry about sounding negative. I don't mind. It's not like this engine is the best thing in the world after sliced bread. It has some strong points and many weak points. 

Overall, this engine is 50% to what you describe, I do some stuff you want, e.g.
* warnings are errors
 * CI has static analysis as one of its steps
 * I have no problem with new c++ version, I use some c++17 stuff here and there, but mostly it's just simple code, which would work even in ancient c++, just because there's no need for variadic templates in most places. But as long as it compiles with all supported compilers, I don't have problem with any specific feature, when necessary. One exception is when it ruins compile time. Very fast build time is one of my main goals.
 * ogl 4.6., dx11, even dx12 (80% finished)
 
but, on the other hand

* I don't follow any guidelines, code is mostly consistent, but there are some old parts, written many years ago when even c++11 was not properly supproted and new parts written recently and they do not match in style
* as you mentioned no code comments or docs

In the end, definitely give [RBFX](https://github.com/rokups/rbfx) a try, it's properly commented and has most stuff you want. Only disadvatage is that its core is build on outdated concepts. It's a fork  or [Urho3D](https://github.com/urho3d/Urho3D) maybe give that also a try. And ofcourse, the most popular Godot Engine.  While you probably won't like its code, it's free oss, easily extendable, with huge community. It might be a bit slower, but 4x games do not require high performance in engine part.

**Comment by metablaster** [comment]
Thank you, your input was very useful, RBFX/Urho3D list of features and design looks very promising, only testing out will give a complete picture.

I wish you all the best! we all need more people like you and fresh innovations because everyone has different expectations and choices are so few.
